### ISSUE TITLE
Write a Java wrapper for the server

### ISSUE THREAD

**Issue by r-jsv** [cthit/VoteIT#5]
With a nice GUI. Could have some of the following features: 
- A list of connected sockets
- A log over incoming answers, connections etc.
- GUI for handling questions, accessCodes, restart, etc.

**Comment by lindskogen** [comment]
Not to complain, but why java? Why not just a web gui?

**Comment by Dr-Horv** [comment]
Spontaniously I can see 2 reasons. 
1. Easier deployment. If anyone should be able to deploy the server (if it's supposed to be run by ´Justerare´) then Java is a good choice since everyone of us use it in our education. And probably more of our students can easier understand the code (i.e. trust it).
2. Typing, easier to be certain of a program being correct if it has proper typing and typechecking.

**Comment by Hultner** [comment]
@Dr-Horv @Ndushi We could easily have both. A Java application with a webui/frontend making it both easily accessible while maintaining the perks of running it on Java (static typing, familiar language to most students, JVM, etc).

**Comment by Dr-Horv** [comment]
@Hultner Indeed. Was what I had in mind. The issue is to have the server-part in Java so that what I answered. 

Thoughts @Ndushi @sveningsonrobin ?

**Comment by Hultner** [comment]
Just took a look at the srouce code. 

The server seems to be written in node so now I kinda agree with @Ndushi. Neither 1 or 2 matters in this case. A malicious party could simply send their own requests hence making static typing useless (if it's only in the frontend). The client part shouldn't really have any of the code that needs trust. Since it's client side any one could easily modify how it interacts with the server (or write a custom client). 

A web client could easily be contained within a single html-file either being served by the node server or locally by opening it in a browser. What really matters in this case would be how the backend would be built. So with the reasoning of @Dr-Horv the server should probably be java but the frontend could be web.

**Comment by Hultner** [comment]
I interpret this issue at it's current state the following way: Write a Java frontend application that communicate with the nodejs backend. @sveningsonrobin: Is this correct?

**Comment by r-jsv** [comment]
The idea is to wrap the server part in java. The frontend should still be based on html.

The reason to wrap the server is because writing commands in terminal/command line etc. takes much longer than using a well designed GUI. If the meeting chairman asks a question, the server administrator must be able to quickly post the question, so that we can collect the answers. And if the question is a form of radiobuttons, checkboxes and inputfields (with default values for vacant, blanks etc.) the process will be much faster.

I suggest that the java wrapper uses the 'Runtime.getRuntime().exec' method to start up the server (http://www.mkyong.com/java/how-to-execute-shell-command-from-java/). So a server admin would do the following:
1. 'javac server.java'
2. 'java server'

Then when pressing the 'start server' button in the GUI, the java wrapper would do:
1. Runtime.getRuntime().exec('npm install');
2. Runtime.getRuntime().exec('node server.js');

The entire functionality could then be built in to this wrapper, which would make the server admin's job much easier.

We can absolutely consider other languages than Java, but as @Dr-Horv said: we all use it in our education so it's easy for anyone to understand.

**Comment by Hultner** [comment]
So basically you want an UI to run npm install and node server.js. An ui that must be launched using two commands?

Wouldn't something like this be easier? 
server-start file:

```
#/usr/bin/env sh
npm install && node server.js
```

So you'd just execute `./server-start` 

Posting questions and such could still be done via an web-interface, no need for that to be done in Java?

**Comment by r-jsv** [comment]
The point with a wrapper is obviosly not to make the server easier to start, but to make the interface more user friendly. Both a web interface and a java-wrapper would do the trick. It's just a matter of personal preference.

**Comment by Hultner** [comment]
But how would the java wrapper make it more user friendly if you still have to start it from the terminal (and compile). Also having node as a prerequisite would probably be the least user friendly task. Have you considered using some sort of node-runtime for the jvm (such as nodyn.io) alleviate this daunting task? 

Then you could probably package everything as a jar and remove the need of a preexisting node runtime environment (that is if the server is supposed to run on ECMA).

**Comment by r-jsv** [comment]
The issue isn't (and have never been) about making it more user friendly to start the server. The part that is supposed to be more user friendly is moving the input from a terminal to a GUI.

**Comment by Hultner** [comment]
That's exactly what I meant, if you need node+jdk and have to run 
`javac someClass.java` and `java someClass` you're not really removing anything from the terminal just adding yet another step. And having a node runtime environment as a requirement makes it even more difficult. You could have a executable (could be "clickable" from ui) that starts the server. After that you could handle everything via an web ui. In this particular case I can't really see how java makes anything simpler, that is unless you've got your ECMA runtime in Java/JVM hence I asked if that's something you've considered.

What I mean is to both minimize the number of steps needed to get it running while also making it as simple as possible to use for novice users. An executable shell script as the example above could be invoked by double-clicking the files, in windows you could make the same thing as a bat script using something like\* 

```
call npm install
call node server.js
```

Which would also be an "double clickable file". I'm not trying to criticize you I'm just asking what you've considered and how it's simplifying things for a novice user (thus making it more user friendly). Neither am I saying that you shouldn't do stuff via an UI I just don't see the need for a _native_ client  and a web client when everything could be handled by the latter. 

*(not sure if that's the exact syntax as I haven't used Windows in many years)

**Comment by r-jsv** [comment]
> if you need node+jdk and have to run 
> javac someClass.java and java someClass you're not really removing anything from the terminal

The point isn't to remove anything from the terminal. That's a whole other issue.

>  just adding yet another step

The number of steps required from a server admin will be just the same (still 2 steps to start the server).

> And having a node runtime environment as a requirement makes it even more difficult

It's safe to say that having node installed is an reasonable requirement before starting a node server.

> I can't really see how java makes anything simpler

As said before. You could use either java or a web UI. Just a matter of personal preference.

> An executable shell script as the example above could be invoked by double-clicking the files, in windows you could make the same thing as a bat script 

Using Java doesn't mean that you can't start the server via a bat/shell script!

**Comment by Hultner** [comment]
> The number of steps required from a server admin will be just the same (still 2 steps to start the server).

No it's one step (double click/execute a single file) versus two+ steps ( compile? execute java file, press button in ui). And if you count in installing node it's even more steps (however something that could either be bundled or installed by an executable if it's not present at the system)

> As said before. You could use either java or a web UI. Just a matter of personal preference.

There's still the need for an web ui if the voting isn't done via the _native_ client, so why not just put it in the same web ui only exposed for the administrators/local connections.

> Using Java doesn't mean that you can't start the server via a bat/shell script!

No it doesn't but why even bother if you've got the web ui up already for the voting?

> The point isn't to remove anything from the terminal. That's a whole other issue.

What's the point then?

**Comment by r-jsv** [comment]
> No it's one step (double click/execute a single file) versus two+ steps ( compile? execute java file, press button in ui). And if you count in installing node it's even more steps (however something that could either be bundled or installed by an executable if it's not present at the system)

Now you are comparing apples and oranges. You can't compare the number of actions it requires for starting a java client via terminal versus how many actions it requires to start a web UI from a script. If we installed node, compiled java and executed java in a script the number of actions compared to using a web UI would be exactly the same (in this case: 1 action)! 

> There's still the need for an web ui if the voting isn't done via the native client, so why not just put it in the same web ui only exposed for the administrators/local connections.

Because that would require a way to securely log in as an administrator, which is not necessary in this case. The easiest way to let an admin control the server is just letting the admin use the host computer. If you read the file 'abstract.md' you will understand that the server will be ran on a computer which is present in the meeting room and that the server admin will interact with the server directly for integrity reasons.

> No it doesn't but why even bother if you've got the web ui up already for the voting?

Because that UI won't be used by the server admin.

> What's the point then?

The point with this issue is to make the process of creating questions, modifying connections, creating accessCodes etc. easier by replacing the terminal input (the commands that the admin gives the server) with a UI used only on the host computer. How many lines of code it takes to start the server is another issue, and surely reducing it to one line of code shouldn't be a problem (and it doesn't depend on wether to use Java or a web UI).

**Comment by Hultner** [comment]
Wouldn't require login if that part of the web ui is only accessible from localhost/loopback. Still don't know if it makes sense to put it in a widely different interface. Seems more sensible too keep all frontend in the same type of codebase as the rest. Another approach could be to hook on an local web view which would be pretty much the same as an web ui that's only locally accessible.

**Comment by Dr-Horv** [comment]
I'm sorry I think I might have created some confusion by my initial comment. And you seem to be arguing on completly different bases.

A. If we ignore the implemented code and things about that for a moment. My thought was that a backend/server would preferably be built in Java so that anyone (even a first year student at study week 6 study period 1) of the IT-program could deploy and admin the server and question and act as 'justerare' and admin of voting during Student Division meetings. To clarify, the goal is to have something any IT-student no matter of coding experience could manage so that not unfair choice of 'justerare'.

Furthermore, Java and all pros with typing, security, JVM etc follows. Certainly there are drawbacks to this, java is old and not as easy to write as perhaps Node but that's another discussion. Personally I like this approach more than having 'justerare' be supervising styrIT etc and it removes any doubt that styrIT is 'jäviga' or cheating etc. I have not had time to discuss this with @sveningsonrobin IRL and he's not up to date with my thoughts and that might be why I confused this discussion, sorry about that.

B. As I now understand it, this issue and @sveningsonrobin is discussing wrapping server and combining that wrapper with a GUI for supplying information to the server (questions posted etc) which at the moment is input by CLI. So, rather a wrapper for CLI -> GUI than easing the deploy process.

**Comment by r-jsv** [comment]
This is not necessary anymore since the terminal input is easy enough to use. Could be implemented in the future if someone finds it to be a necessity.
